### DISCUSSION TITLE
parallel temper move

### DISCUSSION THREAD

**Discussion by directysj** [mlund/faunus#374] [Q&A]
Hello @mlund 

I am testing parallel temper in Faunus.

I have tried to run `temper.yml` script example.
The command is used as -- `$yason.py temper.yml | mpirun --np 3 --stdin all faunus`

It stuck after the command and nothing has happened, no errors. Only creating three files with names `mpi0.stdout`, `mpi1.stdout`, and `mpi2.stdout`. My system (Debian 10) is installed openmpi version 3.1.3 and GCC 10.3.0.

I performed `$make test` after `$make install`, which fails for `temper.yml`.

 Can I ask your suggestions? 

Thanking you in advance!!

**Comment by mlund** [comment]
The parallel temper example must be run through the notebook as `temper.yml` is a template used to generate the three input files for each replica. You'll need to install Jupyter Notebook. Could you describe what is failing with `make test` and `make install`? I have it passing on my machine.

**Reply by directysj** [reply]
I have `Jupyter Notebook` installed as a part of `anaconda 3`. `$make install` does not produce any error.  I am posting errors for failing `temper.yml` below:

```
5/33 Testing: temper
5/33 Test: temper
Command: "/usr/bin/sh" "-c" "jupyter nbconvert --to html --execute temper.ipynb --ExecutePreprocessor.timeout=None"
Directory: /home/user/softwares/mpiFaunus/faunus/examples/temper
"temper" start time: Jun 11 16:51 CEST
Output:
----------------------------------------------------------
[NbConvertApp] Converting notebook temper.ipynb to html
[NbConvertApp] Executing notebook with kernel: python3

CellExecutionError: An error occurred while executing the following cell:
------------------
%%bash
if [[ -z "${FAUNUS_EXECUTABLE}" ]]; then
  mpirun -np 6 faunus -i input.json
else
  echo "Seems we're running CTest - use Faunus target from CMake"
  "${MPIEXEC}" -np 6 "${FAUNUS_EXECUTABLE}" -i input.json --nobar
fi

[0;31mCalledProcessError[0m: Command 'b'if [[ -z "${FAUNUS_EXECUTABLE}" ]]; then\n  mpirun -np 6 faunus -i input.json\nelse\n  echo "Seems we\'re running CTest - use Faunus target from CMake"\n  "${MPIEXEC}" -np 6 "${FAUNUS_EXECUTABLE}" -i input.json --nobar\nfi\n'' returned non-zero exit status 1.
CalledProcessError: Command 'b'if [[ -z "${FAUNUS_EXECUTABLE}" ]]; then\n  mpirun -np 6 faunus -i input.json\nelse\n  echo "Seems we\'re running CTest - use Faunus target from CMake"\n  "${MPIEXEC}" -np 6 "${FAUNUS_EXECUTABLE}" -i input.json --nobar\nfi\n'' returned non-zero exit status 1.

<end of output>
Test time =   3.25 sec
----------------------------------------------------------
Test Failed.
```

**Reply by mlund** [reply]
Difficult to interpret the error, but I think you now have the input files generated (`mpi0.input.json, ...`), so could you try to run `mpirun -np 6 faunus -i input.json` from the terminal in the `temper` directory? You can also open the notebook with `jupyter-notebook temper.ipynb`

**Reply by directysj** [reply]
`$make test` is passed now. I just edited the `temper.ipynb` and changed the number of processes as per the number of threads available in my system, which is less than 6. 

So, `parallel temper` is working only using `input.json` file instead of `.yaml` files?

**Reply by mlund** [reply]
The faunus executable can handle JSON files, only. `yason.py` can be used to convert and pipe in a single yaml file, but for parallel tempering, multiple input files are required (prefixed with `mpiX.`) and piping is therefore not possible.

**Reply by directysj** [reply]
Thank you so much. I was using it in this way and it worked.

**Reply by directysj** [reply]
Hello @mlund 
I would like to ask one query. What is `I `stand for in parallel temper move: `temper: {format: XYZQI}` . I am ommiting `I` from my input file because I would like to swap coordinates and charges between the replicas.

**Reply by directysj** [reply]
Hello @mlund 

While running `Faunus` with `parallel temper` move. I am getting error for one of the process (thread) like shown below: 

`E: sanity: step 95806: peptide23 mass center out of sync by 0.535 Å`

After this message the thread goes into sleeping state while others show running status on `$top` command. 

The `parallel temper` move is used in the input script shown below:

```
moves:
 - rcmc:         {repeat: 2}
 - transrot:     {molecule: peptide, repeat: N}
 - moltransrot:  {molecule: peptide, dp: 0.2, dprot: 0.1, repeat: N}
 - temper:         {format: xyzq}
```

Can you give some suggestions/comments?

Thank you in anticipation!!

**Reply by directysj** [reply]
> Hello @mlund
> 
> While running `Faunus` with `parallel temper` move. I am getting error for one of the process (thread) like shown below:
> 
> `E: sanity: step 95806: peptide23 mass center out of sync by 0.535 Å`
> 
> After this message the thread goes into sleeping state while others show running status on `$top` command.
> 
> The `parallel temper` move is used in the input script shown below:
> 
> ```
> moves:
>  - rcmc:         {repeat: 2}
>  - transrot:     {molecule: peptide, repeat: N}
>  - moltransrot:  {molecule: peptide, dp: 0.2, dprot: 0.1, repeat: N}
>  - temper:         {format: xyzq}
> ```
> 
> Can you give some suggestions/comments?
> 
> Thank you in anticipation!!

this is due to the long polymer's self interactions.

**Reply by mlund** [reply]
Aha, yes that makes sense. The mass center cannot be calculated if the polymers are longer than halt the box length. My suggestion is to make sure the system runs well without tempering, then switch it on. I btw. just updated the error message so that it's a bit more helpful (cde22e540b812ecce9208fe13804a4c8eafd82a1)

**Reply by directysj** [reply]
> Aha, yes that makes sense. The mass center cannot be calculated if the polymers are longer than halt the box length. My suggestion is to make sure the system runs well without tempering, then switch it on. I btw. just updated the error message so that it's a bit more helpful ([cde22e5](https://github.com/mlund/faunus/commit/cde22e540b812ecce9208fe13804a4c8eafd82a1))

Thank you for your suggestion. Now, it is more explanatory.

**Reply by directysj** [reply]
Hello @mlund ,

I have tested `parallel tempering` on `~ 5000 beads` system with `8 `and `12 replicas`. I am showing simulation times below ---

******************** `8 replicas (8 MPI processes)` ********************
```
Increment in temperature within replicas               Simulation time (in minutes)
                          0.20                                                      2224.67
                          0.50                                                      3695.18
                          1.00                                                      5255.50
                          2.00                                                      3298.92
                          3.00                                                      4060.10
```

******************** `12 replicas (12 MPI processes)` ********************

```
Increment in temperature within replicas               Simulation time (in minutes)
                         0.20                                                        4201
                         0.50                                                        4457
                         1.00                                                        3294
                         2.00                                                        3026
```

The same system takes `1803 minutes` in `serial`.  It is true that swapping between replicas differ with increment in temperature within replicas.  Any suggestion(s) to improve the speedup? 

Thank you.

**Reply by mlund** [reply]
I just checked the manual where it says _"the move is performed exactly once per Monte Carlo cycle"_. Can you confirm this in your setup? This may be too often for your system and we could consider implementing a way to do this at a user defined interval. Besides that, I think the main advantage of tempering is that rare events are better sampled in each replica than if they were brute force sampled _N_ times longer.

**Reply by directysj** [reply]
> I just checked the manual where it says _"the move is performed exactly once per Monte Carlo cycle"_. Can you confirm this in your setup? This may be too often for your system and we could consider implementing a way to do this at a user defined interval. Besides that, I think the main advantage of tempering is that rare events are better sampled in each replica than if they were brute force sampled _N_ times longer.

Thank you so much for the quick reply. The `temper move` is performed once a MC cycle. If you can add user defined interval for it then I can test if this make any speedup. I agree with you that replicas sample rare events better.

**Reply by mlund** [reply]
OK, could you try the new branch in PR #381? See the list of changes of how to use it.

**Reply by directysj** [reply]
> OK, could you try the new branch in PR #381? See the list of changes of how to use it.

Thank you so much for adding `nskip` to `temper` move. I will test it and let you know any differences.

**Reply by directysj** [reply]
> OK, could you try the new branch in PR #381? See the list of changes of how to use it.

I have tested for one run and the `speedup` is `21%` corresponding to `skip 100`--> `temper: {format: xyzq, nstep: 100}`
